# Simple Snakemake example

This is a simple Snakemake toy example that maps reads and identifies variants using Ugandan mosquito genome reads, intended for learning how to use the [Snakemake](https://snakemake.readthedocs.io/en/stable/) pipeline management system. 

This example `Snakefile` is based heavily off of (read: essentially plagiarized from) the official Snakemake basic tutorial [found here](https://snakemake.readthedocs.io/en/stable/tutorial/basics.html).

It uses a tiny subset of the genomes of Ugandan mosquitoes generated by the Ag1000G project, [described here](https://www.malariagen.net/data/ag1000g-phase3-snp).

Specifically, it maps the reads near the Vgsc/kdr insecticide-related locus to the AgamP4 genome (using just chromosome arm _2L_) and calls variants.

The example dataset can be generated beforehand with `scripts/generate_sample_dataset.sh`.
